storage.go

```diff
@@ -325,6 +352,68 @@ func (s *Storage) keyList() (map[string]bool, uint64, error) {
 	}
 	return keyList, rev, nil
 }
+
+// keyListAdd adds the given key to the keylist
+func (s *Storage) keyListAdd(key string) error {
+	keyList, keyListRev, err := s.keyList()
+	if err != nil && !errNotFound(err) {
+		return err
+	}
+	if keyList == nil {
+		keyList = make(map[string]bool)
+	}
+	// If the key is already in the keylist there's nothing to do.
+	if _, exists := keyList[key]; exists {
+		return nil
+	}
+	keyList[key] = true
+	bytes, err := json.Marshal(keyList)
+	if err != nil {
+		return errors.AddContext(err, "failed to serialise the new key list")
+	}
+	err = s.SkyDB.Write(bytes, s.KeyListDataKey, keyListRev+1)
+	if err != nil {
+		return errors.AddContext(err, "failed to store the key list")
+	}
+	return nil
+}
+
+// keyListDelete deletes the given key from the keylist
+func (s *Storage) keyListDelete(key string) error {
+	keyList, keyListRev, err := s.keyList()
+	if err != nil && !errNotFound(err) {
+		return err
+	}
+	// If the keylist is empty there's nothing to do.
+	if keyList == nil {
+		return nil
+	}
+	// If the key is not in the keylist there's nothing to do.
+	if _, exists := keyList[key]; !exists {
+		return nil
+	}
+	delete(keyList, key)
+	bytes, err := json.Marshal(keyList)
+	if err != nil {
+		return errors.AddContext(err, "failed to serialise the new key list")
+	}
+	err = s.SkyDB.Write(bytes, s.KeyListDataKey, keyListRev+1)
+	if err != nil {
+		return errors.AddContext(err, "failed to store the key list")
+	}
+	return nil
+}
+
+// errNotFound checks the various failure modes of the registry which all mean
+// that the entry was not found.
+func errNotFound(err error) bool {
+	// This check covers all cases:
```

Comment:
This is a tiny bit confusing cause it's not obvious how those cases are covered.

Reply:
Hm, would it help if I list here their respective error messages or would the verbose example from https://github.com/SkynetLabs/certmagic-storage-skydb/pull/3#discussion_r662169926 be a better approach?

Reply:
Yes, I think so. I don't trust the whole "common substring"-approach because someone might accidentally change the exposed error strings since it's not obvious that we rely on their exact values. So being explicit is probably best.